cell formula refering to DefinedName is not calculated correctly

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Description**

<!--
Briefly describe the problem you are having in a few paragraphs.
-->
Here's a unit test to show the issue:

```golang
func TestCalcCellValueWithDefinedName(t *testing.T) {
	cellData := [][]interface{}{
		{1, 4, nil, "Month", "Team", "Sales"},
	}
	prepareData := func() *File {
		f := NewFile()
		for r, row := range cellData {
			for c, value := range row {
				cell, _ := CoordinatesToCellName(c+1, r+1)
				assert.NoError(t, f.SetCellValue("Sheet1", cell, value))
			}
		}
		assert.NoError(t, f.SetDefinedName(&DefinedName{Name: "ref_to_sheet1_a1", RefersTo: "Sheet1!A1", Scope: "Workbook"}))
		return f
	}
	f := prepareData()
	assert.NoError(t, f.SetCellFormula("Sheet1", "C1", "=ref_to_sheet1_a1"))
	_, err := f.CalcCellValue("Sheet1", "C1")
	assert.NoError(t, err)
}
```

**Describe the results you received:**

```text
=== RUN   TestCalcCellValueWithDefinedName

    TestCalcCellValueWithDefinedName: calc_test.go:817: 
        	Error Trace:	calc_test.go:817
        	Error:      	Received unexpected error:
        	            	#NAME?
        	Test:       	TestCalcCellValueWithDefinedName
--- FAIL: TestCalcCellValueWithDefinedName (0.00s)
FAIL
```

**Describe the results you expected:**

Successful unit test

**Output of `go version`:**

```text
go version go1.14.4 windows/amd64
```

**Excelize version or commit ID:**

```text
Excelize 2.2.0 or commit ID 0aa15106947965bdae9daae7571a4a3f569bf32d
```

**Environment details (OS, Microsoft Excel™ version, physical, etc.):**

Windows 10

--- a/calc.go
+++ b/calc.go
@@ -458,6 +458,12 @@ func isOperatorPrefixToken(token efp.Token) bool {
 func (f *File) parseToken(sheet string, token efp.Token, opdStack, optStack *Stack) error {
 	// parse reference: must reference at here
 	if token.TSubType == efp.TokenSubTypeRange {
+		for _, definedName := range f.GetDefinedName() {
+			if definedName.Name == token.TValue {
+				token.TValue = definedName.RefersTo
+				break
+			}
+		}
 		result, err := f.parseReference(sheet, token.TValue)
 		if err != nil {
 			return errors.New(formulaErrorNAME)

--- a/calc.go
+++ b/calc.go
@@ -453,16 +453,27 @@ func isOperatorPrefixToken(token efp.Token) bool {
 	return false
 }
 
+func (f *File) getDefinedNameRefTo(definedNameName string, currentSheet string) (refTo string) {
+	for _, definedName := range f.GetDefinedName() {
+		if definedName.Name == definedNameName {
+			refTo = definedName.RefersTo
+			// worksheet scope takes precedence over scope workbook when both definedNames exist
+			if definedName.Scope == currentSheet {
+				break
+			}
+		}
+	}
+	return refTo
+}
+
 // parseToken parse basic arithmetic operator priority and evaluate based on
 // operators and operands.
 func (f *File) parseToken(sheet string, token efp.Token, opdStack, optStack *Stack) error {
 	// parse reference: must reference at here
 	if token.TSubType == efp.TokenSubTypeRange {
-		for _, definedName := range f.GetDefinedName() {
-			if definedName.Name == token.TValue {
-				token.TValue = definedName.RefersTo
-				break
-			}
+		refTo := f.getDefinedNameRefTo(token.TValue, sheet)
+		if refTo != "" {
+			token.TValue = refTo
 		}
 		result, err := f.parseReference(sheet, token.TValue)
 		if err != nil {

--- a/calc_test.go
+++ b/calc_test.go
@@ -2,6 +2,7 @@ package excelize
 
 import (
 	"path/filepath"
+	//"strconv"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -790,4 +791,26 @@ func TestCalcCellValue(t *testing.T) {
 	_, err = f.CalcCellValue("Sheet1", "A1")
 	assert.EqualError(t, err, "not support UNSUPPORT function")
 	assert.NoError(t, f.SaveAs(filepath.Join("test", "TestCalcCellValue.xlsx")))
+
+}
+
+func TestCalcCellValueWithDefinedName(t *testing.T) {
+	cellData := [][]interface{}{
+		{1, 4, nil, "Month", "Team", "Sales"},
+	}
+	prepareData := func() *File {
+		f := NewFile()
+		for r, row := range cellData {
+			for c, value := range row {
+				cell, _ := CoordinatesToCellName(c+1, r+1)
+				assert.NoError(t, f.SetCellValue("Sheet1", cell, value))
+			}
+		}
+		assert.NoError(t, f.SetDefinedName(&DefinedName{Name: "ref_to_sheet1_a1", RefersTo: "Sheet1!A1", Scope: "Workbook"}))
+		return f
+	}
+	f := prepareData()
+	assert.NoError(t, f.SetCellFormula("Sheet1", "C1", "=ref_to_sheet1_a1"))
+	_, err := f.CalcCellValue("Sheet1", "C1")
+	assert.NoError(t, err)
 }

--- a/calc_test.go
+++ b/calc_test.go
@@ -795,7 +795,7 @@ func TestCalcCellValue(t *testing.T) {
 
 func TestCalcCellValueWithDefinedName(t *testing.T) {
 	cellData := [][]interface{}{
-		{1, 4, nil, "Month", "Team", "Sales"},
+		{"A1 value", "B1 value", nil},
 	}
 	prepareData := func() *File {
 		f := NewFile()
@@ -805,11 +805,15 @@ func TestCalcCellValueWithDefinedName(t *testing.T) {
 				assert.NoError(t, f.SetCellValue("Sheet1", cell, value))
 			}
 		}
-		assert.NoError(t, f.SetDefinedName(&DefinedName{Name: "ref_to_sheet1_a1", RefersTo: "Sheet1!A1", Scope: "Workbook"}))
+		assert.NoError(t, f.SetDefinedName(&DefinedName{Name: "defined_name1", RefersTo: "Sheet1!A1", Scope: "Workbook"}))
+		assert.NoError(t, f.SetDefinedName(&DefinedName{Name: "defined_name1", RefersTo: "Sheet1!B1", Scope: "Sheet1"}))
+
 		return f
 	}
 	f := prepareData()
-	assert.NoError(t, f.SetCellFormula("Sheet1", "C1", "=ref_to_sheet1_a1"))
-	_, err := f.CalcCellValue("Sheet1", "C1")
+	assert.NoError(t, f.SetCellFormula("Sheet1", "C1", "=defined_name1"))
+	result, err := f.CalcCellValue("Sheet1", "C1")
 	assert.NoError(t, err)
+	// DefinedName with scope WorkSheet takes precedence over DefinedName with scope Workbook, so we should get B1 value
+	assert.Equal(t, "B1 value", result, "=defined_name1")
 }

--- a/calc_test.go
+++ b/calc_test.go
@@ -2,7 +2,6 @@ package excelize
 
 import (
 	"path/filepath"
-	//"strconv"
 	"testing"
 
 	"github.com/stretchr/testify/assert"

--- a/sheet.go
+++ b/sheet.go
@@ -1421,7 +1421,7 @@ func (f *File) GetDefinedName() []DefinedName {
 				RefersTo: dn.Data,
 				Scope:    "Workbook",
 			}
-			if dn.LocalSheetID != nil {
+			if dn.LocalSheetID != nil && *dn.LocalSheetID >= 0 {
 				definedName.Scope = f.getSheetNameByID(*dn.LocalSheetID + 1)
 			}
 			definedNames = append(definedNames, definedName)

